### PR TITLE
Teamwhitelistsupportonclaims

### PR DESCRIPTION
**Pull request by ritmanda**
teamwhitelisting based on claims present userinfo body

### CODE REVIEW

File: pkg/providers/openid/openid.go [hunk 1]
```diff
@@ -57,6 +57,39 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
...
+		m := f.(map[string]interface{})
+		for k := range m {
+			if k == cfg.Cfg.TeamWhiteListClaim {
+				claimval, ok := m[k].(string)
+				if !ok {
+					log.Error("TeamWhiteList claim sent in openID user body cannot be casted as string")
+					// continue auth with existing teammemberships for user
+				}
+
+				user.TeamMemberships = append(user.TeamMemberships, claimval)
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:90]: append only if casting is success

File: handlers/auth.go [hunk 2]
```diff
@@ -139,26 +139,29 @@ func verifyUser(u interface{}) (bool, error) {
...
+		if len(cfg.Cfg.TeamWhiteList) != 0 {
+			for _, team := range user.TeamMemberships {
+				for _, wl := range cfg.Cfg.TeamWhiteList {
+					if team == wl {
+						log.Debugf("verifyUser: Success! found user.TeamWhiteList in TeamWhiteList: %s for user %s", wl, user.Username)
+						return true, nil
+					}
 				}
 			}
+			return false, fmt.Errorf("verifyUser: user.TeamMemberships %s not found in TeamWhiteList: %s for user %s", user.TeamMemberships, cfg.Cfg.TeamWhiteList, user.Username)
```
Comment (kumarkowsikt) [inline handlers/auth.go]: Is this needed when we have below return statement ?

File: pkg/cfg/cfg.go [hunk 3]
```diff
@@ -43,21 +43,22 @@ import (
...
+	SocketMode         int      `mapstructure:"socket_mode"`
+	SocketGroup        string   `mapstructure:"socket_group"`
+	DocumentRoot       string   `mapstructure:"document_root" envconfig:"document_root"`
+	WriteTimeout       int      `mapstructure:"writeTimeout"`
+	ReadTimeout        int      `mapstructure:"readTimeout"`
+	IdleTimeout        int      `mapstructure:"idleTimeout"`
+	Domains            []string `mapstructure:"domains"`
+	WhiteList          []string `mapstructure:"whitelist"`
+	TeamWhiteList      []string `mapstructure:"teamWhitelist"`
+	TeamWhiteListClaim string   `mapstructure:"teamWhitelistclaim"`
```
Comment (kumarkowsikt) [inline pkg/cfg/cfg.go]: Add description on how this attribute is used

File: pkg/providers/openid/openid.go (hunk 1 above)
Comment (ritmanda) [reply pkg/providers/openid/openid.go:90]: fixed

File: handlers/auth.go (hunk 2 above)
Comment (ritmanda) [reply handlers/auth.go]: fixed

File: pkg/providers/openid/openid.go [hunk 4]
```diff
@@ -57,6 +57,41 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
...
+	if err = appendTeamMembershipsFromCustomClaim(data, user); err != nil {
+		log.Error(err)
+		return err
+	}
+
 	user.PrepareUserData()
 	return nil
 }
+
+// appendTeamMembershipsFromCustomClaim appends teammembership values in user If
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go]: if

File: handlers/auth.go [hunk 5]
```diff
@@ -139,26 +139,29 @@ func verifyUser(u interface{}) (bool, error) {
...
-		return false, fmt.Errorf("verifyUser: user.Username not found in WhiteList: %s", user.Username)
-
-	// TeamWhiteList
-	case len(cfg.Cfg.TeamWhiteList) != 0:
-		for _, team := range user.TeamMemberships {
-			for _, wl := range cfg.Cfg.TeamWhiteList {
-				if team == wl {
-					log.Debugf("verifyUser: Success! found user.TeamWhiteList in TeamWhiteList: %s for user %s", wl, user.Username)
-					return true, nil
+		log.Info("User not in userwhitelist, checking in teamWhiteList")
```
Comment (kumarkowsikt) [inline handlers/auth.go]: change this to debug

**Review by kumarkowsikt — APPROVED** [review]

File: pkg/cfg/cfg.go (hunk 3 above)
Comment (ritmanda) [reply pkg/cfg/cfg.go]: addressed

File: pkg/providers/openid/openid.go (hunk 4 above)
Comment (ritmanda) [reply pkg/providers/openid/openid.go]: addressed

File: handlers/auth.go (hunk 5 above)
Comment (ritmanda) [reply handlers/auth.go]: addressed

**Pull request merged**